aptos-move/framework/aptos-framework/sources/confidential_asset/confidential_asset.move

```diff
@@ -0,0 +1,1127 @@
+        proof: RotationProof) acquires ConfidentialAssetStore
+    {
+        let user = signer::address_of(sender);
+        let current_ek = encryption_key(user, token);
+
+        let ca_store = borrow_global_mut<ConfidentialAssetStore>(get_user_address(user, token));
+
+        let pending_balance = confidential_balance::decompress_balance(&ca_store.pending_balance);
+
+        // We need to ensure that the pending balance is zero before rotating the key.
+        // To guarantee this, the user must call `rollover_pending_balance_and_freeze` beforehand.
+        assert!(confidential_balance::is_zero_balance(&pending_balance), error::invalid_state(ENOT_ZERO_BALANCE));
+
+        let current_balance = confidential_balance::decompress_balance(&ca_store.actual_balance);
+
+        confidential_proof::verify_rotation_proof(&current_ek, &new_ek, &current_balance, &new_balance, &proof);
+
+        ca_store.ek = new_ek;
+        // We don't need to update the pending balance here, as it has been asserted to be zero.
+        ca_store.actual_balance = confidential_balance::compress_balance(&new_balance);
+        ca_store.normalized = true;
+    }
+
+    /// Implementation of the `normalize` entry function.
+    public fun normalize_internal(
+        sender: &signer,
+        token: Object<Metadata>,
+        new_balance: confidential_balance::ConfidentialBalance,
+        proof: NormalizationProof) acquires ConfidentialAssetStore
+    {
+        let user = signer::address_of(sender);
+        let sender_ek = encryption_key(user, token);
+
+        let ca_store = borrow_global_mut<ConfidentialAssetStore>(get_user_address(user, token));
+
+        assert!(!ca_store.normalized, error::invalid_state(EALREADY_NORMALIZED));
+
+        let current_balance = confidential_balance::decompress_balance(&ca_store.actual_balance);
+
+        confidential_proof::verify_normalization_proof(&sender_ek, &current_balance, &new_balance, &proof);
+
+        ca_store.actual_balance = confidential_balance::compress_balance(&new_balance);
+        ca_store.normalized = true;
+    }
+
+    /// Implementation of the `rollover_pending_balance` entry function.
+    public fun rollover_pending_balance_internal(
+        sender: &signer,
+        token: Object<Metadata>) acquires ConfidentialAssetStore
+    {
+        let user = signer::address_of(sender);
+
+        assert!(has_confidential_asset_store(user, token), error::not_found(ECA_STORE_NOT_PUBLISHED));
+
+        let ca_store = borrow_global_mut<ConfidentialAssetStore>(get_user_address(user, token));
+
+        assert!(ca_store.normalized, error::invalid_state(ENORMALIZATION_REQUIRED));
+
+        let actual_balance = confidential_balance::decompress_balance(&ca_store.actual_balance);
+        let pending_balance = confidential_balance::decompress_balance(&ca_store.pending_balance);
+
+        confidential_balance::add_balances_mut(&mut actual_balance, &pending_balance);
```

Comment:
and actual balance overflow (2^128-1 -> 1) is impossible here because FA total supply is 128-bit?

Reply:
serious FA deployers will set a maximum_supply, correct? @gregnazario or @lightmark